src/script/lua_api/l_object.cpp

```diff
@@ -668,68 +585,57 @@ int ObjectRef::l_get_bone_position(lua_State *L)
 int ObjectRef::l_set_attach(lua_State *L)
 {
 	GET_ENV_PTR;
-
 	ObjectRef *ref = checkobject(L, 1);
 	ObjectRef *parent_ref = checkobject(L, 2);
-	ServerActiveObject *co = getobject(ref);
+	ServerActiveObject *sao = getobject(ref);
 	ServerActiveObject *parent = getobject(parent_ref);
-	if (co == NULL)
+	if (sao == nullptr || parent == nullptr)
 		return 0;
-
-	if (parent == NULL)
-		return 0;
-
-	if (co == parent)
+	if (sao == parent)
 		throw LuaError("ObjectRef::set_attach: attaching object to itself is not allowed.");
 
-	// Do it
 	int parent_id = 0;
 	std::string bone;
 	v3f position = v3f(0, 0, 0);
 	v3f rotation = v3f(0, 0, 0);
 	bool force_visible;
-	co->getAttachment(&parent_id, &bone, &position, &rotation, &force_visible);
+
+	sao->getAttachment(&parent_id, &bone, &position, &rotation, &force_visible);
 	if (parent_id) {
 		ServerActiveObject *old_parent = env->getActiveObject(parent_id);
-		old_parent->removeAttachmentChild(co->getId());
+		old_parent->removeAttachmentChild(sao->getId());
 	}
 
-	bone = "";
-	if (!lua_isnil(L, 3))
-		bone = readParam<std::string>(L, 3);
-	position = v3f(0, 0, 0);
-	if (!lua_isnil(L, 4))
-		position = read_v3f(L, 4);
-	rotation = v3f(0, 0, 0);
-	if (!lua_isnil(L, 5))
-		rotation = read_v3f(L, 5);
+	bone      = lua_isnil(L, 3) ? ""           : readParam<std::string>(L, 3);
```

Comment:
How about `bone = readParam<std::string>(L, 3, "")` for a default return value?

Sometime in the future (not in this PR), `readParam` should be made comprehensive and work for all sorts of types, e.g. s32, v3s16, and suchlike, all with default values options.  That way, we don't have to have such a mixture of `readParam` and `luaL_check<type>` and `read_<type>` with `nil` and type checks all over the place.

Reply:
nice idea about the standard